src/festim/coupled_heat_hydrogen_problem.py

```diff
@@ -0,0 +1,183 @@
+        if isinstance(
+            value,
+            HTransportProblemDiscontinuous
+            | HTransportProblemPenalty
+            | HydrogenTransportProblemDiscontinuousChangeVar,
+        ):
+            raise NotImplementedError(
+                "Coupled heat transfer - hydorgen transport simulations with "
+                "HydrogenTransportProblemDiscontinuousChangeVar, "
+                "HTransportProblemPenalty or"
+                "HydrogenTransportProblemDiscontinuousChangeVar, "
+                "not currently supported"
+            )
+        elif not isinstance(value, HydrogenTransportProblem):
+            raise TypeError(
+                "hydrogen_problem must be a festim.HydrogenTransportProblem object"
+            )
+        self._hydrogen_problem = value
+
+    @property
+    def non_matching_meshes(self):
+        return self.heat_problem.mesh.mesh != self.hydrogen_problem.mesh.mesh
+
+    def initialise(self):
+        # make sure both problems have the same initial time step and final time,
+        # use minimal initial value of the two and maximal final time of the two
+        min_initial_dt = min(
+            self.heat_problem.settings.stepsize.initial_value,
+            self.hydrogen_problem.settings.stepsize.initial_value,
+        )
+        self.heat_problem.settings.stepsize.initial_value = min_initial_dt
+        self.hydrogen_problem.settings.stepsize.initial_value = min_initial_dt
+
+        if (
+            self.heat_problem.settings.final_time
+            != self.hydrogen_problem.settings.final_time
+        ):
+            raise ValueError(
+                "Final time values in the heat transfer and hydrogen transport "
+                "model must be the same"
+            )
+
+        self.heat_problem.initialise()
+
+        self.heat_problem.show_progress_bar = False
+
+        if self.non_matching_meshes:
+            V = fem.functionspace(self.hydrogen_problem.mesh.mesh, ("P", 1))
+            T_func = fem.Function(V)
+
+            nmm_interpolate(T_func, self.heat_problem.u)
+
+            self.hydrogen_problem.temperature = T_func
+        else:
+            self.hydrogen_problem.temperature = self.heat_problem.u
+
+        self.hydrogen_problem.initialise()
+
+    def iterate(self):
+        self.heat_problem.iterate()
+
+        if self.non_matching_meshes:
+            nmm_interpolate(
+                self.hydrogen_problem.temperature_fenics, self.heat_problem.u
+            )
+
+        self.hydrogen_problem.iterate()
+
+        # use the same time step for both problems, use minimum of the two
+        next_dt_value = min(
+            float(self.hydrogen_problem.dt), float(self.heat_problem.dt)
+        )
+        self.heat_problem.dt = as_fenics_constant(
+            value=next_dt_value, mesh=self.heat_problem.mesh.mesh
+        )
+        self.hydrogen_problem.dt = as_fenics_constant(
+            value=next_dt_value, mesh=self.hydrogen_problem.mesh.mesh
+        )
```

Comment:
Oh no this won't work! it needs to be:
```suggestion
        # use the same time step for both problems, use minimum of the two
        next_dt_value = min(
            float(self.hydrogen_problem.dt), float(self.heat_problem.dt)
        )
        self.heat_problem.dt.value = next_dt_value
        self.hydrogen_problem.dt.value = next_dt_value
```

Reply:
Consider this test:

```python
test_heat_problem = F.HeatTransferProblem(
    mesh=test_mesh,
    subdomains=test_subdomains,
    settings=F.Settings(atol=1, rtol=1, transient=True, stepsize=0.5, final_time=5),
)

test_hydrogen_problem = F.HydrogenTransportProblem(
    mesh=test_mesh,
    subdomains=test_subdomains,
    species=[test_H],
    settings=F.Settings(atol=1, rtol=1, transient=True, stepsize=1.5, final_time=5),
)

test_coupled_problem = F.CoupledTransientHeatTransferHydrogenTransport(
    heat_problem=test_heat_problem,
    hydrogen_problem=test_hydrogen_problem,
)

test_coupled_problem.initialise()

dt_hydrogen_used_in_formulation = test_hydrogen_problem.dt

# run
test_coupled_problem.iterate()
dt_hydrogen_object_after_iterate = test_hydrogen_problem.dt

# test


assert dt_hydrogen_used_in_formulation == dt_hydrogen_object_after_iterate  # should fail


```

Reply:
added test